include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,413 @@
+        __atomic_flag.store(__full_status, sycl::memory_order::release);
+    }
+
+    template <typename _Subgroup, typename _BinaryOp>
+    _T
+    cooperative_lookback(const _Subgroup& __subgroup, _BinaryOp __binary_op)
+    {
+        _T __running = oneapi::dpl::unseq_backend::__known_identity<_BinaryOp, _T>;
+        auto __local_id = __subgroup.get_local_id();
+
+        for (int __tile = static_cast<int>(__tile_id) - 1; __tile >= 0; __tile -= SUBGROUP_SIZE)
+        {
+            _AtomicFlagT __tile_flag_atomic(*(__flags_begin + __tile + __padding - __local_id));
+            _T __tile_flag = __initialized_status;
+
+            // Load flag from a previous tile based on my local id.
+            // Spin until every work-item in this subgroup reads a valid status
+            do
+            {
+                __tile_flag = __tile_flag_atomic.load(sycl::memory_order::acquire);
+            } while (!sycl::all_of_group(__subgroup, __tile_flag != __initialized_status));
+
+            bool __is_full = __tile_flag == __full_status;
+            auto __is_full_ballot = sycl::ext::oneapi::group_ballot(__subgroup, __is_full);
+            ::std::uint32_t __is_full_ballot_bits{};
+            __is_full_ballot.extract_bits(__is_full_ballot_bits);
+
+            _AtomicValueT __tile_value_atomic(
+                *((__is_full ? __full_vals_begin : __partial_vals_begin) + __tile + __padding - __local_id));
+            _T __tile_val = __tile_value_atomic.load(sycl::memory_order::acquire);
+
+            auto __lowest_item_with_full = sycl::ctz(__is_full_ballot_bits);
+            _T __contribution = __local_id <= __lowest_item_with_full
+                                    ? __tile_val
+                                    : oneapi::dpl::unseq_backend::__known_identity<_BinaryOp, _T>;
+
+            // Running reduction of all of the partial results from the tiles found, as well as the full contribution from the closest tile (if any)
+            __running = __binary_op(__running, sycl::reduce_over_group(__subgroup, __contribution, __binary_op));
+
+            // If we found a full value, we can stop looking at previous tiles. Otherwise,
+            // keep going through tiles until we either find a full tile or we've completely
+            // recomputed the prefix using partial values
+            if (__is_full_ballot_bits)
+                break;
+        }
+        return __running;
+    }
+
+    const uint32_t __tile_id;
+    _FlagType* __flags_begin;
+    _T* __full_vals_begin;
+    _T* __partial_vals_begin;
+    _AtomicFlagT __atomic_flag;
+    _AtomicValueT __atomic_partial_value;
+    _AtomicValueT __atomic_full_value;
+};
+
+template <typename _FlagType, typename _Type, typename _BinaryOp, typename _KernelName>
+struct __lookback_init_submitter;
+
+template <typename _FlagType, typename _Type, typename _BinaryOp, typename... _Name>
+struct __lookback_init_submitter<_FlagType, _Type, _BinaryOp,
```

Comment:
It is a bit complicated because we need to initialize the flags to different values based on whether it is out-of-bounds (padding) or not. We could possibly create a buffer on the host and copy it over, but I haven't done any testing to see if there is a performance difference from copying data from the host vs initializing it on the device itself with the kernel.

Reply:
Yeah, I think if it requires extra host-> device copy it will be worse.  

My hope was to be able to handle this logic with a subset of threads in the main kernel, and use default constructed "0" to mean "uninitialized".  Certainly not a blocker for this PR but its maybe worth investigating to see if it is possible (and a benefit) in a follow up PR.

Reply:
Lets not worry about this within the context of this PR.